File: maude-report.r
library(httr)
library(jsonlite)
library(dplyr)

# Take User Input
product_code_input <- readline(prompt = "Enter the Product Code:")
start_date_input <- readline(prompt = "Start Date [YYYY-MM-DD]:")
end_date_input <- readline(prompt = "End Date [YYYY-MM-DD]:")

# Construct Query
product_code <- paste("device.device_report_product_code.exact:", product_code_input, sep="")
date_received <- paste("date_received:[",start_date_input,"+TO+",end_date_input,"]", sep="")
separator <- "+AND+"
limit <- "&limit=500"  # Increased limit to get more data

base_endpoint <- "https://api.fda.gov/device/event.json?search="
field_term <- paste(date_received, separator, product_code, sep = "")
query <- paste0(base_endpoint, field_term, limit)

# Fetch Data from OpenFDA
response <- GET(query)

if (http_status(response)$category == "Success") {
    # Parse JSON response
    data <- fromJSON(content(response, "text"), flatten = TRUE)$results
    
    if (length(data) == 0) {
        stop("No results found.")
    }

    # Function to Extract Nested Data
    safe_extract <- function(data, column) {
        if (column %in% names(data)) {
            return(data[[column]])
        } else if (grepl("\\.", column)) {
            # Handle nested columns
            parts <- strsplit(column, "\\.")[[1]]
            if (all(parts %in% names(data))) {
                return(data[[parts[1]]][[parts[2]]])
            }
        }
        return(rep(NA, nrow(data)))
    }

    # Extract Brand Name
    extract_brand_name <- function(device_list) {
        sapply(device_list, function(device) {
            if (is.data.frame(device) && "brand_name" %in% names(device)) {
                return(device$brand_name[1])
            } else {
                return(NA)
            }
        })
    }

    # Create Resultant Data Frame 
    combined_data <- data.frame(
        Report_Number = safe_extract(data, "report_number"),
        Brand_Name = extract_brand_name(data$device),
        Event_Type = I(safe_extract(data, "event_type")),
        Product_Problems = I(safe_extract(data, "product_problems")),
        stringsAsFactors = FALSE
    )

    # Function to unnest list columns
    unnest_column <- function(df, col) {
        unnested <- lapply(df[[col]], function(x) if(length(x) == 0) NA else paste(x, collapse = "; "))
        df[[col]] <- unlist(unnested)
        return(df)
    }

    # Unnest Event_Type and Product_Problems columns
    combined_data <- unnest_column(combined_data, "Event_Type")
    combined_data <- unnest_column(combined_data, "Product_Problems")

    # Brand Name and Event Type Report
    event_report <- combined_data %>%
        group_by(Brand_Name, Event_Type) %>%
        summarise(Event_Count = n(), .groups = 'drop') %>%
        arrange(Brand_Name, desc(Event_Count))

    # Brand Name and Product Problems Report
    problem_report <- combined_data %>%
        group_by(Brand_Name, Product_Problems) %>%
        summarise(Problem_Count = n(), .groups = 'drop') %>%
        arrange(Brand_Name, desc(Problem_Count))

    # Write CSV files
    tryCatch({
        write.csv(event_report, "processed/event_type_report.csv", row.names = FALSE)
        cat("Brand Name and Event Type Report generated: event_type_report.csv\n")
    }, error = function(e) {
        cat("Error writing event_type_report.csv:", conditionMessage(e), "\n")
    })

    tryCatch({
        write.csv(problem_report, "processed/product_problems_report.csv", row.names = FALSE)
        cat("Brand Name and Product Problems Report generated: product_problems_report.csv\n")
    }, error = function(e) {
        cat("Error writing product_problems_report.csv:", conditionMessage(e), "\n")
    })

} else {
    stop("Failed to retrieve data: ", http_status(response)$message)
}

File: maude_report.py
import requests
import pandas as pd
import os
import json

# Take user input
def user_input():
    product_code_input = input("Enter the Product Code: ")
    start_date_input = input("Start Date [YYYY-MM-DD]: ")
    end_date_input = input("End Date [YYYY-MM-DD]: ")
    return product_code_input, start_date_input, end_date_input

# Construct the API query
def query_constructor(start_date, end_date, product_code):
    base_endpoint = "https://api.fda.gov/device/event.json?search="
    separator = "+AND+"
    limit = "&limit=500"
    
    date_received = f"date_received:[{start_date}+TO+{end_date}]"
    p_code = f"device.device_report_product_code.exact:{product_code}"
    
    query = f"{base_endpoint}{date_received}{separator}{p_code}{limit}"
    return query

# Fetch data from openFDA
def fetch_API(query):
    response = requests.get(query)
    if response.status_code == 200:
        data = response.json().get('results', [])
        if data:
            return data
        else:
            raise ValueError("No results found.")
    else:
        raise Exception(f"Failed to retrieve data: {response.status_code} {response.text}")

# Normalize the fetched data
def normalize_data(data):
    combined_data = pd.json_normalize(data)

    combined_data['Brand_Name'] = combined_data['device'].apply(
        lambda x: x[0]['brand_name'] if isinstance(x, list) and x else None
    )
    combined_data['Event_Type'] = combined_data['event_type'].apply(
        lambda x: x if pd.notnull(x) else None
    )
    combined_data['Product_Problems'] = combined_data['product_problems'].apply(
        lambda x: '; '.join(x) if isinstance(x, list) and x else None
    )

    return combined_data

# Create event report
def create_event_report(combined_data):
    event_report = (combined_data
        .groupby(['Brand_Name', 'Event_Type'])
        .size()
        .reset_index(name='Event_Count')
        .sort_values(by=['Brand_Name', 'Event_Count'], ascending=[True, False])
    )
    return event_report

# Create problem report
def create_problem_report(combined_data):
    problem_report = (combined_data
        .groupby(['Brand_Name', 'Product_Problems'])
        .size()
        .reset_index(name='Problem_Count')
        .sort_values(by=['Brand_Name', 'Problem_Count'], ascending=[True, False])
    )
    return problem_report

# Save report to CSV
def save_report_to_csv(report, filename):
    os.makedirs('reports', exist_ok=True)
    try:
        report.to_csv(f"reports/{filename}", index=False)
        print(f"Report generated: {filename}")
    except Exception as e:
        print(f"Error writing {filename}: {str(e)}")

# Save JSON response to file
def save_json(data):
    os.makedirs('reports', exist_ok=True)
    try:
        with open("reports/fda_data.json", "w") as json_file:
            json.dump(data, json_file, indent=4)
        print("JSON data saved as fda_data.json")
    except Exception as e:
        print(f"Error saving JSON data: {str(e)}")

# Save normalized data to JSON file
def save_normalized_json(data):
    os.makedirs('reports', exist_ok=True)
    try:
        normalized_json_file_path = "reports/normalized_data.json"
        data.to_json(normalized_json_file_path, orient='records', lines=True)
        print(f"Normalized data saved as normalized_data.json")
    except Exception as e:
        print(f"Error saving normalized data: {str(e)}")

def main():
    try:
        product_code, start_date, end_date = user_input()
        query = query_constructor(start_date, end_date, product_code)
        data = fetch_API(query)
        save_json(data)  # Save the raw JSON data
        
        combined_data = normalize_data(data)  # Normalize the data
        event_report = create_event_report(combined_data)  # Create event report
        problem_report = create_problem_report(combined_data)  # Create problem report

        save_report_to_csv(event_report, "event_type_report.csv")  # Save event report
        save_report_to_csv(problem_report, "product_problems_report.csv")  # Save problem report
        
        save_normalized_json(combined_data)  # Save normalized data
    except Exception as e:
        print(f"An error occurred: {str(e)}")

if __name__ == "__main__":
    main()
